Leave signals alone

gelisam opened the pull request:
Fixes #51

mvdan commented:
Did you write multiple commits on purpose, or do you mean to rebase/squash this before merging?

mvdan commented on src/Control/Monad/Ghc.hs:
```diff
@@ -25,12 +27,22 @@ instance (Functor m, Monad m) => Applicative (GhcT m) where
   pure  = return
   (<*>) = ap
 
+-- adapted from https://github.com/ghc/ghc/blob/59fa7b32b018a91f81773ca676251a0b2761ef56/compiler/main/GHC.hs#L454-L463
```
Would it be possible to use a tag for the link here? Would make the link a bit more human-friendly and shorter.

We then do depend on then not modifying tags, but I doubt they will. And if they do, it's unlikely that the line numbers will be wrong anyway.

gelisam replied:
done

mvdan commented on src/Control/Monad/Ghc.hs:
```diff
@@ -25,12 +27,22 @@ instance (Functor m, Monad m) => Applicative (GhcT m) where
+rawRunGhcT mb_top_dir ghct = do
+  ref <- liftIO $ newIORef (error "empty session")
+  let session = GHC.Session ref
+  flip GHC.unGhcT session $ {-GHC.withSignalHandlers $-} do -- do _not_ catch ^C
```
I assume `{-GHC.withSignalHandlers $-}` is to make it very clear what our modification is?

gelisam replied:
yes

mvdan commented on unit-tests/run-unit-tests.hs:
```diff
@@ -221,6 +227,47 @@ test_normalize_type = TestCase "normalize_type" [mod_file] $ do
+-- note that hint was _not_ overwriting the signal handlers when the hint interpreter
+-- was itself executed inside the ghci interpreter. for this reason, this test always
+-- succeeds when executed from ghci and ghcid, regardless of whether the problematic
+-- behaviour has been fixed or not.
```
Perhaps make the test fail (or at least show a clear "skipped test because ...") under ghci? Assuming that is possible.

gelisam replied:
I don't know how to determine whether we're running under ghci or not. `GHC.runGhcT` must somehow detect it in order to not install the signal handlers, but how this is accomplished is not obvious from the code.

mvdan replied:
Not a big deal - was just wondering if this was easy.

mvdan commented:
Also, please briefly summarize your changes in the commit description, like mentioning the added exported name, and the changes that this fix will do to people's use of hint.

I assume you also considered exporting a new name instead of breaking backwards compatibility slightly. I would likely not introduce a new name either, as you explained how the old behavior wasn't really useful to anyone. In any case, please do mention that too in the commit message.

gelisam commented:
> Did you write multiple commits on purpose

I did, but now that you mention it, I do remember that you always ask contributors to squash their commits. Maybe you should add a `CONTRIBUTING.md` file pointing out this policy, so you don't have to repeat it every time?

I have now squashed my commits.

mvdan commented:
:) I just prefer avoiding redundant commits. But of course, they can be useful to separate different pieces of work, which is why I was asking.

I don't mention it in the contribution guide because GitHub now has a web interface to squash easily, so I don't really have to depend on others doing it. But I still ask, in case they did mean to separate their work in multiple commits.

gelisam commented:
> please briefly summarize your changes in the commit description

I have now modified my commit description to include this.

> like mentioning the added exported name

I did not export any new name, `rawRunGhcT` is private to `Control.Monad.Ghc`

> I assume you also considered exporting a new name instead of breaking backwards compatibility slightly.

I was hoping that my intention to modify the existing function was clear in #51, when I said we should change the _default_ behaviour. I expected you might ask for a separate function then, but you didn't, so I went ahead and modified the existing function's behaviour.

mvdan commented:
> I did not export any new name

Err, I meant to delete that part of my comment.

> I was hoping that my intention to modify the existing function was clear

Yes - it is clear to me, but I was talking more in terms of someone reading the commit message and not having the full context. Thanks for updating the message.

Happy to merge once you've addressed the code comments.

gelisam commented:
> In any case, please do mention [breaking backwards compatibility slightly] too in the commit message.

I have now modified my commit message to include this as well.

mvdan commented:
Happy to merge once CI is green. Seems like the tests don't build properly, perhaps the deps aren't set up right.

gelisam commented:
Let's see...

The ghc-7.10.3 build fails with `Not in scope: ‘GHC.withCleanupSession’. I am guessing that the problem is that the `runGhcT` code I have copied is from ghc HEAD, but this code has changed over the past few years. Indeed:

* only [ghc-8.2 calls `withSignalHandlers`](https://github.com/ghc/ghc/blob/ghc-8.2/compiler/main/GHC.hs#L434-L442)
* [ghc-8.0 calls `installSignalHandlers`](https://github.com/ghc/ghc/blob/ghc-8.0/compiler/main/GHC.hs#L436-L445) instead, which does _not_ restore the handlers afterwards
* [ghc-7.10 does not call either](https://github.com/ghc/ghc/blob/ghc-7.10/compiler/main/GHC.hs#L436-L445), nor does it call `withCleanupSession` (I don't know what that cleans). I remember reading some ghc notes saying that the ghc API was no longer automatically installing signal handlers and that users now had to call `installSignalHandlers` if they want them, so it might not be possible to prevent `runInterpreter` from installing signal handlers when using ghc-7.10 :(

ghc-8.4 will roll out soon, and your policy is to only support the last three major ghc versions, so can we just drop support for ghc-7.10?

mvdan commented:
I adopted the three-version policy very recently, simply to keep the project easier to maintain in the long run. I did not imagine supporting more than three versions would be a problem this soon :)

Dropping 7.10 is fine, as long as you add 8.4, which is already present as a 8.4.0 git build in hvrs repo: https://launchpad.net/~hvr/+archive/ubuntu/ghc/+index?batch=75&memo=75&start=75

I also understand that the added code should run fine on 8.0, 8.2, and 8.4 without any CPP magic? Either way, please have a look at the failures on those versions.

Feel free to leave the dynamic build+test one with 8.2, the latest stable version.

mvdan commented:
To add to my comment above - I was planning on doing another Hint release sometime this spring, so I can simply make it line up with the GHC 8.4 release. That Hint release will add support for 8.4, and drop 7.8 and 7.10 (as per the new policy).

gelisam commented:
I have now modified the `.travis.yml` file, and I have updated the commit message to reflect the change.

mvdan commented:
Travis is still very unhappy :)

mvdan commented:
It might make sense to do the addition of the 8.4.0 git build in a separate pull request, as it's likely that it will require a few fixes here and there. At least, it should be its own commit.

gelisam commented:
> At least, it should be its own commit.

See, this is why squashing commits is a bad idea! Now I have to manually separate the changes related to ghc-8.4 from the rest of my changes, whereas if the commits were separate, it would have been trivial to rebase cherry-pick it onto master :)

I will extract the changes (fortunately only the `.travis.yml` file changed so this will be easy) and create another PR.

mvdan commented:
Well, I never asked that you squash all changes :) I simply asked if you wanted to keep the commits separate or not. The commits you initially posted seemed like amends to previous commits, which is why I wondered if you meant to squash them before merging.

mvdan commented:
@gelisam please rebase - master now supports GHC 8.4 and dropped 7.10.

gelisam commented:
I have now rebased.
